easybuild/framework/easyblock.py

```diff
@@ -1914,15 +1922,15 @@ def install_extensions_sequential(self, install=True):
                 ext.toolchain.prepare(onlymod=self.cfg['onlytcmod'], silent=True, loadmod=False,
                                       rpath_filter_dirs=self.rpath_filter_dirs)
 
-            # real work
+            # actual installation of the extension
             if install:
                 try:
-                    ext.prerun()
+                    ext.pre_install_extension()
```

Comment:
This seems to make sense, but it's not actually this simple: if an easyblock (say `PythonPackage`) hasn't been adjusted to use `pre_install_extension` rather than `prerun`, then it breaks things.

For example, when installing `matplotlib`:

<details>

```
$ eb matplotlib-3.7.2-gfbf-2023a.eb -f
Installing 'kiwisolver' extension (3/5) ━━━━━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━ 0:00:02
EasyBuild crashed! Please consider reporting a bug, this should not happen...

Traceback (most recent call last):
  File "/usr/lib64/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib64/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/main.py", line 768, in <module>
    main_with_hooks()
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/main.py", line 754, in main_with_hooks
    main(args=args, prepared_cfg_data=(init_session_state, eb_go, cfg_settings))
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/main.py", line 715, in main
    hooks, do_build)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/main.py", line 552, in process_eb_args
    exit_on_failure=exit_on_failure)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/main.py", line 173, in build_and_install_software
    raise ec_res['err']
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/main.py", line 134, in build_and_install_software
    (ec_res['success'], app_log, err) = build_and_install_one(ec, init_env)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/easyblock.py", line 4258, in build_and_install_one
    result = app.run_all_steps(run_test_cases=run_test_cases)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/easyblock.py", line 4132, in run_all_steps
    self.run_step(step_name, step_methods)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/easyblock.py", line 3967, in run_step
    step_method(self)()
  File "/arcanine/scratch/gent/vo/000/gvo00002/vsc40023/easybuild/eb5/easybuild-easyblocks/easybuild/easyblocks/generic/pythonbundle.py", line 125, in extensions_step
    super(PythonBundle, self).extensions_step(*args, **kwargs)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/easyblock.py", line 2894, in extensions_step
    self.install_all_extensions(install=install)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/easyblock.py", line 1880, in install_all_extensions
    self.install_extensions_sequential(install=install)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/easyblock.py", line 1930, in install_extensions_sequential
    txt = ext.install_extension()
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/extensioneasyblock.py", line 155, in install_extension
    EasyBlock.patch_step(self, beginpath=self.ext_dir)
  File "/user/gent/400/vsc40023/easybuild/eb5/easybuild-framework/easybuild/framework/easyblock.py", line 2598, in patch_step
    beginpath = self.src[srcind]['finalpath']
TypeError: string indices must be integers
```

</details>

By not calling `prerun` anymore, the work that's implemented by `PythonPackage.prerun` is not being done anymore...

So we need to be a bit smarter here, and ensure that `prerun` is still run if the easyblock does not implement `pre_install_extension` yet...

Reply:
Understood, the issue is that the main installation function in the base `EasyBlock` class has to explicitly call any deprecated methods (`prerun`, `run`, ...) found in the child EasyBlock used by the extension (if any). And that obviously won't happen if it only calls the newer methods (`pre_install_extension`, `extension_install`, ...), which probably do not exist on older extension EasyBlocks.

That was indeed a bit trickier to fix than expected. Main problem is that the base `Extension` class has to define the deprecated methods for as long as they are allowed to be used because the deprecated methods in the child EasyBlock might call the methods from the parent `Extension` class with `super`. However, it is not trivial to detect if a child class overloads one of the deprecated methods from the `Extension` class.

So I added a new method to `Extension` as a wrapper to all calls to `pre_install_extension`, `extension_install`, etc that will detect the presence of any deprecated method in the extension EasyBlock and use that one if present.